proxy/coco/src/peer/run_state.rs

```diff
@@ -0,0 +1,383 @@
+            // Go from [`Input::Stopped`] to [`Input::Offline`] once we are listening.
+            (Status::Stopped(_since), Event::Protocol(ProtocolEvent::Listening(_addr))) => {
+                self.status = Status::Started(Instant::now());
+
+                vec![]
+            },
+            // Sync with first incoming peer.
+            //
+            // In case the peer is configured to sync on startup we start syncing, otherwise we go
+            // online straight away.
+            // TODO(xla): Also issue sync if we come online after a certain period of being
+            // disconnected from any peer.
+            (Status::Started(_since), Event::Protocol(ProtocolEvent::Connected(ref peer_id))) => {
+                self.connected_peers.insert(peer_id.clone());
+
+                if self.config.sync_on_startup {
+                    self.status = Status::Syncing(Instant::now(), 1);
+
+                    vec![
+                        Command::SyncPeer(peer_id.clone()),
+                        Command::StartSyncTimeout(self.config.sync_period),
+                    ]
+                } else {
+                    self.status = Status::Online(Instant::now());
+
+                    vec![]
+                }
+            },
+            // Sync until configured maximum of peers is reached.
+            (Status::Syncing(since, syncs), Event::Protocol(ProtocolEvent::Connected(peer_id)))
+                if *syncs < self.config.sync_max_peers =>
+            {
+                self.connected_peers.insert(peer_id.clone());
+                if syncs + 1 == self.config.sync_max_peers {
+                    self.status = Status::Online(Instant::now());
+                } else {
+                    self.status = Status::Syncing(*since, syncs + 1);
+                }
+
+                vec![Command::SyncPeer(peer_id)]
+            }
+            // Go online if we exceed the sync period.
+            (Status::Syncing(_since, _syncs), Event::Timeout(TimeoutEvent::SyncPeriod)) => {
+                self.status = Status::Online(Instant::now());
+
+                vec![]
+            },
+            // Remove peer that just disconnected.
+            (_, Event::Protocol(ProtocolEvent::Disconnecting(peer_id))) => {
+                self.connected_peers.remove(&peer_id);
+
+                // Go offline if we have no more connected peers left.
+                if self.connected_peers.is_empty() {
+                    self.status = Status::Offline(Instant::now());
+                }
+
+                vec![]
+            },
+            // Announce new updates while the peer is online.
+            (
+                Status::Online(_) | Status::Started(_) | Status::Syncing(_, _),
+                Event::Announce(AnnounceEvent::Tick),
+            ) => vec![Command::Announce],
+            _ => vec![],
+        }
+    }
+}
+
+#[allow(clippy::needless_update, clippy::panic, clippy::unwrap_used)]
+#[cfg(test)]
+mod test {
```

Comment:
Are there any tests for going to the `Offline` state? Or in general, do you think there are missing tests for state transitions?

Also, tests for events that come in but we're not in the happy path state?

I've used `cargo-tarpaulin` in the past to see which code paths have I tested for getting code coverage.

Reply:
Good catch, trusty mate tarpaulin indeed revealed that this case wasn't covered. For future generations #939.